src/pipeline.rs

```diff
@@ -21,66 +30,120 @@ pub fn build_graphs(
+                                    row_s.send(row).unwrap();
+                                })
+                            }
+                        });
                     }
-                });
+                    FileType::Tsv => {
+                        let config_col_num = config.columns.len();
+                        s.spawn(move |_| {
+                            for input in files_r {
+                                read_file(input, config.log_every_n as u64, |line| {
+                                    let row = parse_tsv_line(line);
+                                    let line_col_num = row.len();
+                                    if line_col_num == config_col_num {
+                                        row_s.send(row).unwrap();
+                                    } else {
+                                        warn!("Wrong number of columns (expected: {}, provided: {}). The line [{}] is skipped.", config_col_num, line_col_num, line);
+                                    }
+                                });
+                            }
+                        });
+                    }
+                }
             }
         }
-    }
 
-    drop(bus);
+        #[allow(clippy::needless_collect)]  // Must collect because of 'drop' calls below
+        let buffers_handles: Vec<_> = (0..processing_worker_num).map(|_| {
+            let hyperedges_r = hyperedges_r.clone();
+            let entity_processor = EntityProcessor::new(
+                config,
+                in_memory_entity_mapping_persistor.clone(),
+            );
+            let sparse_matrices = sparse_matrices_desc.clone();
 
-    let mut sparse_matrices = vec![];
-    for join_handle in sparse_matrix_threads {
-        let sparse_matrix = join_handle
-            .join()
-            .expect("Couldn't join on the associated thread");
-        sparse_matrices.push(sparse_matrix);
+            s.spawn(move |_| {
+                let mut buffers: Vec<_> = sparse_matrices.iter().map(|smd| smd.make_buffer()).collect();
+
+                for row in hyperedges_r {
+                    let row: Vec<SmallVec<[_; SMALL_VECTOR_SIZE]>> = row;
+                    let iterator = entity_processor.process_row_and_get_edges(&row);
+
+                    for buffer in &mut buffers {
+                        for hashes in iterator.clone() {
+                            buffer.handle_pair(&hashes);
+                        }
+                    }
+                }
+
+                buffers
+            })
+        }).collect();
+
+        // Drop it so all channels are closed when work is finished
+        drop(hyperedges_s);
+        drop(hyperedges_r);
+
+        buffers_handles.into_iter().map(|h| h.join().expect("Thread finished job")).collect()
+    }).expect("Threads finished work");
+
+    let merging_time = Instant::now();
+
+    let mut buffers_transposed = Vec::with_capacity(matrices_n);
```

Comment:
Extract some functions from here (transpose_buffers, finish_build

Reply:
Extracted transpose. 

About extracing other stuff like worker spawning. Experimented with it and didnt like it.

Methods would have huge headers. Got some issues with lifetimes (we would pass scope there)

'Pipeline' function is like a glue code that glues channels/threads. And I kinda lika that all that glue is in one place and not scattered around.